streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -103,32 +109,35 @@ void createTasks(final Collection<TopicPartition> assignment) {
             throw new IllegalStateException(logPrefix + "consumer has not been initialized while adding stream tasks. This should not happen.");
         }
 
-        // do this first as we may have suspended standby tasks that
-        // will become active or vice versa
-        standby.closeNonAssignedSuspendedTasks(assignedStandbyTasks);
-        active.closeNonAssignedSuspendedTasks(assignedActiveTasks);
-
-        addStreamTasks(assignment);
-        addStandbyTasks();
+        if (!assignment.isEmpty()) {
+            addActiveTasks(assignment);
+        }
+        if (!addedStandbyTasks.isEmpty()) {
+            addNewStandbyTasks(addedStandbyTasks);
+        }
         // Pause all the partitions until the underlying state store is ready for all the active tasks.
         log.trace("Pausing partitions: {}", assignment);
         consumer.pause(assignment);
     }
 
-    private void addStreamTasks(final Collection<TopicPartition> assignment) {
+    private void addActiveTasks(final Collection<TopicPartition> assignment) {
         if (assignedActiveTasks == null || assignedActiveTasks.isEmpty()) {
             return;
         }
-        final Map<TaskId, Set<TopicPartition>> newTasks = new HashMap<>();
-        // collect newly assigned tasks and reopen re-assigned tasks
-        log.debug("Adding assigned tasks as active: {}", assignedActiveTasks);
+
+        final Map<TaskId, Set<TopicPartition>> newTasks = new HashMap<>(addedActiveTasks);
         for (final Map.Entry<TaskId, Set<TopicPartition>> entry : assignedActiveTasks.entrySet()) {
             final TaskId taskId = entry.getKey();
             final Set<TopicPartition> partitions = entry.getValue();
 
             if (assignment.containsAll(partitions)) {
                 try {
-                    if (!active.maybeResumeSuspendedTask(taskId, partitions)) {
+                    if (active.maybeResumeSuspendedTask(taskId, partitions)) {
```

Comment:
This comment is for line 148: the `assignment` is passed to stream twice, where `assignor#onAssignment` is called first in which we've already updated assigned / added / revoked tasks; and then `listener#onPartitionsAssigned` is called again, therefore the condition in line 133 should always be true, otherwise it should be a bug right?

In this case, we would not need the `assignment ` passed in `addActiveTasks` again; instead we can just do the following in order in the `listener#onPartitionsAssigned`

1) loop over `revoked active / standby` and close them.
2) loop over `(assigned - added) active` and resume them.
3) loop over `added active / standby` to create-as-new.

Reply:
I agree it should always be true in eager case, but I was thinking it could be used in eager to skip over the tasks in `assignedActiveTasks` which are not actually new (of course the warning logged does not make sense in that case)

That's effectively what is happening now, except 2. and 3. are squashed together (mostly because we may fail to resume suspended tasks for valid reasons, ie if  the partitions of the task change, in which case we should add to `newTasks` and recreate)
Ideally we could avoid looping through all assigned tasks during cooperative, possibly by leveraging `active.suspendedTaskIds`

Reply:
Also, the assignment passed to `onPartitionsAssigned` is not the same as what's passed in to `onAssignment`, which is given the "whole" assignment vs (assigned - owned) for `onPartitionsAssigned`. So we can calculate `addedActiveTasks` in `onAssignment` as the tasks we did not previously have, but we can leverage (assigned - owned) to get all the tasks we need to resume

Reply:
That's a good point that the same taskID may change with different partitions.

I think what I'm trying to explore is whether we can make the logic a bit clearer to read while being protocol agnostic. That is, we can either 

1) reason about the source-of-truth for added / revoked / remained tasks for both active and standby in `assignor#onAssignment`, and then we do not need to rely on the parameters passed in the `listener#onPartitionsRevoked / Assigned` at all.

2) rely on the `listener#onPartitionsRevoked / Assigned` to get the list of revoked / remained / added tasks.

Reply:
The way I see it, the two callbacks have related but still complimentary information -- only `onAssignment` will have the full assignment, whereas only `onPartitionsAssigned` tells you specifically which partitions/tasks must be (re)initialized so you can ignore the protocol.

Without using this we would need to loop through all `assignedActiveTasks`, resuming the suspended and recreating any that were closed because that failed, or they were still restoring or not yet running. ie we'd have to keep track of which tasks/partitions need to be reinitialized anyway, which is exactly what we get for free from the listener's callback. True there is a workaround for `onPartitionsAssigned`, but we definitely need `onPartitionsRevoked` because we don't even know the new assignment when its called.

Anyways, I do agree it is somewhat confusing but I just wanted to share my thoughts